code/modules/research/server.dm

```diff
@@ -75,123 +121,88 @@
 	refresh_working()
 
 /obj/machinery/rnd/server/proc/mine()
-	. = base_mining_income
-	var/penalty = max((get_env_temp() - temp_tolerance_high), 0) * temp_penalty_coefficient
-	current_temp = get_env_temp()
-	. = max(. - penalty, 0)
+	// Cheap way to refresh if we are operational or not.  mine() is run on the tech web
+	// subprocess.  This saves us having to run our own subprocess
+	refresh_working()
+	if(working)
+		var/penalty = max((get_env_temp() - temp_tolerance_high), 0) * temp_penalty_coefficient
+		return list(TECHWEB_POINT_TYPE_GENERIC = max(base_mining_income - penalty, 0))
+	else
+		return list(TECHWEB_POINT_TYPE_GENERIC = 0)
+
 
 /obj/machinery/rnd/server/proc/get_env_temp()
-	var/turf/L = loc
-	if(isturf(L))
-		return L.temperature
-	return 0
-
-/obj/machinery/rnd/server/proc/produce_heat(heat_amt)
-	if(!(machine_stat & (NOPOWER|BROKEN))) //Blatently stolen from space heater.
-		var/turf/L = loc
-		if(istype(L))
-			var/datum/gas_mixture/env = L.return_air()
-			if(env.temperature < (heat_amt+T0C))
-
-				var/transfer_moles = 0.25 * env.total_moles()
-
-				var/datum/gas_mixture/removed = env.remove(transfer_moles)
-
-				if(removed)
-
-					var/heat_capacity = removed.heat_capacity()
-					if(heat_capacity == 0 || heat_capacity == null)
-						heat_capacity = 1
-					removed.temperature = min((removed.temperature*heat_capacity + heating_power)/heat_capacity, 1000)
-
-				env.merge(removed)
-				air_update_turf()
-
-/proc/fix_noid_research_servers()
-	var/list/no_id_servers = list()
-	var/list/server_ids = list()
-	for(var/obj/machinery/rnd/server/S in GLOB.machines)
-		switch(S.server_id)
-			if(-1)
-				continue
-			if(0)
-				no_id_servers += S
-			else
-				server_ids += S.server_id
-
-	for(var/obj/machinery/rnd/server/S in no_id_servers)
-		var/num = 1
-		while(!S.server_id)
-			if(num in server_ids)
-				num++
-			else
-				S.server_id = num
-				server_ids += num
-		no_id_servers -= S
+	return thermo.get_env_temp()
```

Comment:
this should be done through the component, not through referencing it.

Reply:
Its it cost effective to keep calling GetComponent for it though?  Whole reason I am deref is the component constantly updates the current temp though its process so I would of thought it be more efficient to just pull that data from that.

Reply:
the thing is that you shouldn't be touching the component at all; the idea of components on TG is that they are generic behavior you can add to an object without having to interact with the component outside of making callbacks.

Everything should be handled through signals (And those signals should not be used like proc calls; they should essentialy be events the component is listening for like for example when the object its attached to is hit)